src/app/components/hooks/use-model-state.ts

```diff
@@ -95,19 +97,20 @@ export const useModelState = <IModelInputState, IModelOutputState>(
         // Don't let client update finished run.
         return oldState;
       }
-      const newState = [...oldState];
+      const newState = [...oldState.map((run) => { return {...run};})];
```

Comment:
Looking at this function and its context, I wouldn't expect this line. This function updates input state of the last run. Why would we make a shallow copy of all the previous runs? This can trigger a few more re-renders than we actually need. I bet it won't be visible in practice and won't cause any real-world performance issues. But more importantly - this might hide some other problem by forcing re-renders. 

Why did you add that and what difference it makes? If it fixes some bug,  maybe it can/should be fixed somewhere else.

Reply:
Hi Piotr, I initially added this as it seemed to solve a bug I was running into integrating setInteractiveState into the app. However I just took a look at it again and you're right, it's not necessary. Thank you for catching it :)

Reply:
So what was the piece that finally solved the issue we're looking at that day?

Reply:
In terms of the issue we were looking at where interactiveState was initially loading as null and then being overwritten by the local state, it was solved by using getInteractiveState() to access interactiveState.